External/FEXCore/Source/Interface/Core/JIT/Arm64/VectorOps.cpp

```diff
@@ -3776,37 +3778,204 @@ DEF_OP(VSShrS) {
+        case 2:
+          return IsHigh128 ? DestIdx >= 12 : DestIdx >= 4;
+        case 4:
+          return IsHigh128 ? DestIdx >= 6 : DestIdx >= 2;
+        case 8:
+          return IsHigh128 ? DestIdx >= 2 : DestIdx >= 1;
+        default:
+          return false;
+      }
+    }();
+
+    // Now sanitize the index so we can cleanly operate
+    // on the four individual 64-bit values
+    const auto SanitizedDstIdx = [DestIdx, ElementSize, IsHigh64, IsHigh128] {
+      uint32_t NewIdx = DestIdx;
+      switch (ElementSize) {
+        case 1:
+          if (IsHigh128) {
+            NewIdx -= 16;
+          }
+          if (IsHigh64) {
+            NewIdx -= 8;
+          }
+          break;
+        case 2:
+          if (IsHigh128) {
+            NewIdx -= 8;
+          }
+          if (IsHigh64) {
+            NewIdx -= 4;
+          }
+          break;
+        case 4:
+          if (IsHigh128) {
+            NewIdx -= 4;
+          }
+          if (IsHigh64) {
+            NewIdx -= 2;
+          }
+          break;
+        case 8:
+          if (IsHigh128) {
+            NewIdx -= 2;
+          }
+          if (IsHigh64) {
+            NewIdx -= 1;
+          }
+          break;
+      }
+      return NewIdx;
+    }();
+    const auto SanitizedDstOffset = SanitizedDstIdx * ElementSizeBits;
+
+    // Now lets create/update our to-be mask
+    const auto UpdateMask = [&] {
+      const uint64_t Mask = (1ULL << ElementSizeBits) - 1;
+      const uint64_t NewEntry = ElementSize == 8 ? 0
+                                                 : (UINT64_MAX & ~(Mask << SanitizedDstOffset));
+
+      auto& Value = IsHigh128 ? Data[1] : Data[0];
+      if (IsHigh64) {
+        Value.UpdateValue(NewEntry, UINT64_MAX);
+      } else {
+        Value.UpdateValue(UINT64_MAX, NewEntry);
+      }
+    };
+
+    // Update our data that we'll load from memory.
+    UpdateMask();
+
+    // Broadcast our source value across a temporary.
+    // This will mask the particular register element
+    // into the final result.
+    switch (ElementSize) {
+      case 1: {
+        dup(VTMP2.Z().VnB(), SrcVector.Z().VnB(), SrcIdx);
+        break;
+      }
+      case 2: {
+        dup(VTMP2.Z().VnH(), SrcVector.Z().VnH(), SrcIdx);
+        break;
+      }
+      case 4: {
+        dup(VTMP2.Z().VnS(), SrcVector.Z().VnS(), SrcIdx);
+        break;
+      }
+      case 8: {
+        dup(VTMP2.Z().VnD(), SrcVector.Z().VnD(), SrcIdx);
+        break;
+      }
+      default:
+        LOGMAN_MSG_A_FMT("Unknown Element Size: {}", ElementSize);
+        return;
     }
-    case 4: {
-      mov(reg.V4S(), Op->DestIdx, GetSrc(Op->SrcVector.ID()).V4S(), Op->SrcIdx);
-    break;
+
+    // Load our mask.
+    aarch64::Label DataLocation;
+    adr(TMP1, &DataLocation);
+    ld1b(VTMP3.Z().VnB(), PRED_TMP_32B.Zeroing(), SVEMemOperand(TMP1));
+
+    // Now finally combine the value from the source temporary we broadcasted into.
```

Comment:
I'm curious if Predicated `CPY` and Predicated `SEL` could be used to make this a bit nicer or not?
Would require generating temporary predicate literals instead of a 256-bit mask.

Reply:
Looks like EXT also is very nice to allow you to barrel rotate any element in to the scalar position for predicated CPY to pick up?

Reply:
Yeah, converting it over to setting up a predicate register greatly simplifies everything